From faa3f7ce8d50bc4c611f2a253fc876e4c700df76 Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@siol.net>
Date: Thu, 16 Aug 2018 20:33:57 +0200
Subject: [PATCH 044/153] clk: sunxi-ng: Consider limits for recalc callback

Currently, if NM PLL is set to unstable rate before Linux kernel is
loaded, it still can be cosidered as valid rate. This can happen if PLL
rate is multiply of wanted rate.

For example, if PLL is set to 128 MHz (below 192 MHz threshold) and
target rate is 32 MHz, clk framework will happily use that.

To counteract this, return 0 as currently set frequency if recalc_rate
function if actuall PLL frequency is outside of working (stable) range.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
[Icenowy: fix a use-before-check error and change returned 1 to 0]
Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/clk/sunxi-ng/ccu_nm.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/clk/sunxi-ng/ccu_nm.c b/drivers/clk/sunxi-ng/ccu_nm.c
index 6fe3c14f7b2d..40dce1aecd25 100644
--- a/drivers/clk/sunxi-ng/ccu_nm.c
+++ b/drivers/clk/sunxi-ng/ccu_nm.c
@@ -102,6 +102,12 @@ static unsigned long ccu_nm_recalc_rate(struct clk_hw *hw,
 	else
 		rate = parent_rate * n / m;
 
+	if (rate < nm->min_rate)
+		rate = 0;
+
+	if (nm->max_rate && rate > nm->max_rate)
+		rate = 0;
+
 	if (nm->common.features & CCU_FEATURE_FIXED_POSTDIV)
 		rate /= nm->fixed_post_div;
 
-- 
2.18.1

